### PR TITLE
Features/thanh/net 8

### PR DESCRIPTION
**Pull request by hoangthanh28**
Hi Erik.
This PR is mainly about upgrading the .net version to latest lts of .net (.net 8). 
I'm not a big fan of floating package version, so please consider this PR as refactor it using [central package management](https://devblogs.microsoft.com/nuget/introducing-central-package-management/) 
Adding some rules check to improve the quality of the code.

Side note: I'm using VS code, the code format a little bit different than your, if you want to reformat using Rider, do let me know, happy to change it.

### CODE REVIEW

File: Directory.Build.props [hunk 1]
```diff
@@ -0,0 +1,47 @@
+<Project>
+
+  <PropertyGroup>
+    <LangVersion>preview</LangVersion>
+    <EnableNETAnalyzers>true</EnableNETAnalyzers>
+    <AnalysisLevel>preview</AnalysisLevel>
+    <NoWarn>$(NoWarn);CS1591;NU5118;NU5128</NoWarn>
+    <TreatWarningsAsErrors>true</TreatWarningsAsErrors>
```
Comment (erikbra) [inline Directory.Build.props:8]: <3 it! If we can keep warnings at 0, that makes the code base better to work with.

File: Directory.Build.props [hunk 2]
```diff
@@ -0,0 +1,47 @@
...
+    <Nullable>enable</Nullable>
+    <ImplicitUsings>enable</ImplicitUsings>
+    <DebugSymbols>true</DebugSymbols>
+    <DebugType>portable</DebugType>
+    <SymbolPackageFormat>snupkg</SymbolPackageFormat>
+    <CodeAnalysisRuleset>$(MSBuildThisFileDirectory)CodeAnalysis.ruleset</CodeAnalysisRuleset>
+    <EnableXlfLocalization>false</EnableXlfLocalization>
```
Comment (erikbra) [inline Directory.Build.props:15]: There are some properties here that I am not familiar with. Have you introduced them on purpose, or are they generated by a tool, without you having any special thought about them? I'm thinking of:

* EnableXlfLocalization
* AccelerateBuildsInVisualStudio

Do these make sense if we don't publish grate as a traditional NuGet library (e.g. for people to reference in their code)?

* ProduceReferenceAssembly
* SymbolPackageFormat

File: Directory.Packages.props [hunk 3]
```diff
@@ -0,0 +1,36 @@
...
+                                          █▀▀██ █ █ ██ ▄▄▄███ █████ ▄▄ █▀▀█ ▀▄ ██
+                                          █▄▄██ ██▄ ██ ▀▀▀███ █████ ▀▀ █▄▄█ ▀▀ ██
+                                          ▀▀▀▀▀▀▀▀▀▀▀▀▀▀▀▀▀▀▀▀▀▀▀▀▀▀▀▀▀▀▀▀▀▀▀▀▀▀▀
+  -->
+
+  <ItemGroup>
+    <!--
+      .net sdk
+    -->
+    <PackageVersion Include="Microsoft.Extensions.Logging.Console"                                       Version="8.0.0" />
```
Comment (erikbra) [inline Directory.Packages.props:14]: I kind-of like using .* on minor versions. It makes it less work-intensive, to upgrade every time a 3rd party has a patch version upgrade.

Do you have other specific thoughts about this? Do you see any drawbacks of using `"8.0.*"` on these dependencies?

**Comment by erikbra** [comment]
Thanks a lot for your PR, @hoangthanh28 ! This looks very good to me. I agree that using Directory.Build.props is a good idea, to avoid having to manage the nuget package versions in every .csproj (even though they are in the same solution). And, I think the IDE support for Directory.Build.props is widespread enough now as well, so I don't see any reason not to start using it :)

And, of course, upgrading to stable .net 8 has been on my todo-list since it was released, I just haven't got around to it yet, so I appreciate your contribution!

Also, a very good idea to add a Codeanalysis.ruleset file. That's a good start! And, if we find out that we want to change any of the rules going forward, it can always be adjusted later.

I had a couple of questions/comments, could you please look through them? Otherwise, I'm good to merge this!

File: grate/Infrastructure/Npgsql/ReflectionNpgsqlQueryParser.cs [hunk 4]
```diff
@@ -20,24 +20,26 @@ public static class ReflectionNpgsqlQueryParser
...
         var npgsqlBatchCommand = typeof(NpgsqlBatchCommand);
 
-        ParseRawQuery = sqlQueryParserType.GetMethod("ParseRawQuery", BindingFlags.Instance | BindingFlags.NonPublic, new [] {npgsqlCommand, typeof(bool), typeof(bool)} )!;
-        
+        ParseRawQuery = sqlQueryParserType.GetMethod("ParseRawQuery", BindingFlags.Instance | BindingFlags.NonPublic, new[] { npgsqlCommand, typeof(bool), typeof(bool) })!;
+
         InternalBatchCommands = npgsqlCommand.GetProperty("InternalBatchCommands", BindingFlags.Instance | BindingFlags.NonPublic)!;
         State = npgsqlCommand.GetProperty("State", BindingFlags.Instance | BindingFlags.NonPublic)!;
-        IsCached = npgsqlCommand.GetProperty("IsCached", BindingFlags.Instance | BindingFlags.NonPublic)!;
+        IsCached = npgsqlCommand.GetProperty("IsCacheable", BindingFlags.Instance | BindingFlags.NonPublic)!;
```
[the remark this answers is not in the document]
Comment (erikbra) [reply grate/Infrastructure/Npgsql/ReflectionNpgsqlQueryParser.cs:39]: Yes, it has (I double-checked myself :) )

**Review by erikbra — APPROVED** [review]
I'm happy to merge this now. Thanks a lot for the PR!

File: grate/grate.csproj [hunk 5]
```diff
@@ -2,36 +2,23 @@
 
   <PropertyGroup>
     <OutputType>Exe</OutputType>
```
[the remark this answers is not in the document]
Comment (aeons) [reply grate/grate.csproj:4]: Yes, you can set them in the Directory.Build.props once and it will be inherited.

**Pull request merged**